### PR TITLE
[AIR] <Part 2> Support metric logging and checkpointing for LightningTrainer

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

There will be a list of PRs for PyTorch Lightning Integration. This is the second one.

- LightningTrainer + Test
- Logging and Checkpointing + Test
- Predictor + Batch Prediction + Test
- Integrate with Tune + Test

Content for this PR:
- [x] Implemented RayModelCheckpoint callback to report latest metrics and checkpoint to trainer session
- [x] Implemented LightningCheckpoint, a subclass of TorchCheckpoint.  
- [x] CI Tests

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/lightning/_lightning_utils.py [hunk 1]
```diff
@@ -46,11 +52,13 @@ def node_rank(self) -> int:
...
     def set_world_size(self, size: int) -> None:
-        logger.warning("world_size setter is disabled in AIR LightningTrainer.")
+        if self.global_rank() == 0:
+            logger.warning("world_size setter is disabled in AIR LightningTrainer.")
         pass
 
     def set_global_rank(self, rank: int) -> None:
-        logger.warning("global_rank setter is disabled in AIR LightningTrainer.")
+        if self.global_rank() == 0:
+            logger.warning("global_rank setter is disabled in AIR LightningTrainer.")
```
Comment (amogkam) [inline python/ray/train/lightning/_lightning_utils.py:61]: are these warnings necessary? is it possible for the user to have access to the environment?

File: python/ray/train/lightning/_lightning_utils.py [hunk 2]
```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
...
+        filename: Optional[str] = None,
+        ver: Optional[int] = None,
+    ) -> str:
+        """
+        Change checkpoint files path to align with AIR checkpoint format.
+
+        e.g. './epoch=2-loss=0.12.ckpt' -> './epoch=2-loss=0.12.ckpt/model'
+        """
+        filepath = super().format_checkpoint_name(metrics, filename, ver)
+        return f"{filepath}/{MODEL_KEY}"
```
Comment (amogkam) [inline python/ray/train/lightning/_lightning_utils.py:134]: do we need to add this MODEL_KEY here?

File: python/ray/train/lightning/_lightning_utils.py [hunk 3]
```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
...
+            return
+        self.is_checkpoint_step = False
+
+        # Report latest logged metrics
+        kwargs = {}
+        metrics = {}
+        for k, v in self._monitor_candidates(trainer).items():
+            if k == "_stage":
+                logger.warning(
+                    "'_stage' is a reserved key in AIR report metrics. "
```
Comment (amogkam) [inline python/ray/train/lightning/_lightning_utils.py:150]: why is this reserved?

File: python/ray/train/lightning/_lightning_utils.py [hunk 4]
```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
...
+    ) -> str:
+        """
+        Change checkpoint files path to align with AIR checkpoint format.
+
+        e.g. './epoch=2-loss=0.12.ckpt' -> './epoch=2-loss=0.12.ckpt/model'
+        """
+        filepath = super().format_checkpoint_name(metrics, filename, ver)
+        return f"{filepath}/{MODEL_KEY}"
+
+    def _session_report(self, trainer: "pl.Trainer", stage: str):
```
Comment (amogkam) [inline python/ray/train/lightning/_lightning_utils.py:136]: add docstring to this please?

File: python/ray/train/lightning/_lightning_utils.py [hunk 5]
```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
...
+
+        metrics["_stage"] = stage
+        kwargs["metrics"] = metrics
+
+        filepath = None
+        if self.monitor:
+            # Capture metric-based top-k checkpoint
+            new_checkpoint = self.best_k_models.keys() - self.last_best_k_models.keys()
+            if new_checkpoint:
+                filepath = new_checkpoint.pop()
```
Comment (amogkam) [inline python/ray/train/lightning/_lightning_utils.py:165]: what is the logic that's going on here? can we always just report the latest checkpoint and use the top k logic from AIR?

File: python/ray/train/lightning/lightning_checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        assert os.path.exists(path), f"Lightning checkpoint {path} doesn't exists!"
+
+        cache_dir = tempfile.mkdtemp()
+        new_checkpoint_path = os.path.join(cache_dir, MODEL_KEY)
+        shutil.copy(path, new_checkpoint_path)
+        if preprocessor:
+            save_preprocessor_to_dir(preprocessor, cache_dir)
+        checkpoint = cls.from_directory(cache_dir)
+        checkpoint._cache_dir = cache_dir
```
Comment (amogkam) [inline python/ray/train/lightning/lightning_checkpoint.py:61]: do we need this line?

File: python/ray/train/tests/test_lightning_trainer.py [hunk 7]
```diff
@@ -164,7 +213,14 @@ def test_trainer_with_categorical_ray_data(ray_start_6_cpus_2_gpus, accelerator)
...
 
-    trainer.fit()
+    results = trainer.fit()
+    assert results.metrics["epoch"] == num_epochs - 1
+    assert (
+        results.metrics["step"] == num_epochs * dataset_size / num_workers / batch_size
+    )
+    assert "loss" in results.metrics
+    assert "val_loss" in results.metrics
+    assert results.checkpoint
```
Comment (amogkam) [inline python/ray/train/tests/test_lightning_trainer.py:223]: can you also add a test for resume_from_checkpoint?

File: python/ray/train/lightning/_lightning_utils.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:134]: Yes, this is for checkpoint loading. For any AIR LightningCheckpoint, we make sure the checkpoint file saves in the format `{local_dir}/MODEL_KEY`, so that we always load from that path, just like what HugginfaceCheckpoint does.

File: python/ray/train/lightning/_lightning_utils.py (hunk 3 above)
Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:150]: We are inserting the `stage` tag ("train_batch_end", "train_epoch_end", "validation_end") into metric dict, we choose "_stage" as a reserved key.

File: python/ray/train/lightning/lightning_checkpoint.py (hunk 6 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_checkpoint.py:61]: We used `tempfile.mkdtemp()` and have to clear this folder afterwards. This cache_dir is deleted in `__del__()` method.

File: python/ray/train/lightning/lightning_trainer.py [hunk 8]
```diff
@@ -325,6 +336,36 @@ def __init__(
...
+            no monitor + no save_top_k:  default = 1, only save last one
+            no monitor +    save_top_k:  if save_top_k != 1, save all
+               monitor + no save_top_k:  default = 1, save the best checkpoint
+               monitor +    save_top_k:  n/a
+        """
+
+        if not lightning_checkpoint_config:
+            lightning_checkpoint_config = {}
+
+        mode = lightning_checkpoint_config.get("mode", "min")
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:356]: Should we be setting `min` as a default? I feel like this should be required if they set a monitor.

File: python/ray/train/lightning/lightning_trainer.py [hunk 9]
```diff
@@ -78,7 +80,6 @@ def __init__(self) -> None:
         self._trainer_fit_params = {}
         self._ddp_strategy_config = {}
         self._model_checkpoint_config = {}
-        self._ray_dataset_iter_config = {}
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:81]: What's this?

File: python/ray/train/lightning/lightning_trainer.py [hunk 10]
```diff
@@ -396,21 +438,12 @@ def _lightning_train_loop_per_worker(config):
...
-                "LightningTrainer will create a Ray ModelCheckpoint callback "
-                "based on `LightningConfig.model_checkpoint_config`. All the other"
-                " ModelCheckpoint callbacks are ignored."
-            )
-        else:
-            callbacks.append(callback)
-    trainer_config["callbacks"] = callbacks
+    # AIR needs a RayModelCheckpoint for metircs logging anyway.
+    trainer_config["enable_checkpointing"] = True
+    os.environ["TUNE_DISABLE_STRICT_METRIC_CHECKING"] = "1"
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:443]: Hmmm, is it possible to set this with a `runtime_env` instead? Seems kind of hard though... This will propagate the env variable to future code that runs in the same script. What do other people think?

**Review by justinvyu — COMMENTED** [review]
Nice! I have a few questions:

Also, do we need to implement a custom `restore` method, or is using `DataParallelTrainer`'s default one ok? The only PTL-specific that you may need to re-specify is `lightning_config`. I.e. re-specifying the lightning module class. Can talk offline about this.

File: python/ray/train/lightning/_lightning_utils.py [hunk 11]
```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
...
+            if trainer.global_rank == 0:
+                kwargs["checkpoint"] = LightningCheckpoint.from_directory(
+                    path=os.path.dirname(filepath)
+                )
+            else:
+                kwargs["checkpoint"] = LightningCheckpoint.from_dict(
+                    {"rank": session.get_world_rank()}
+                )
+
+        self.last_best_k_models = deepcopy(self.best_k_models)
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:184]: Actually `self.best_k_models` are just filenames.

File: python/ray/train/lightning/lightning_trainer.py [hunk 12]
```diff
@@ -308,6 +309,16 @@ def __init__(
         preprocessor: Optional[Preprocessor] = None,
         resume_from_checkpoint: Optional[Checkpoint] = None,
     ):
+        if not lightning_config:
+            lightning_config = LightningConfigBuilder().build()
+
+        if not run_config:
+            run_config = RunConfig()
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_trainer.py:316]: 👌 I saw the TorchTrainer uses `if not` clause. But yes, let's use `or` to make it more pythonic.

File: python/ray/train/lightning/_lightning_utils.py [hunk 13]
```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
...
+            if k == "_stage":
+                logger.warning(
+                    "'_stage' is a reserved key in AIR report metrics. "
+                    "Original values are overwritten!"
+                )
+                continue
+            if isinstance(v, torch.Tensor):
+                metrics[k] = v.item()
+
+        metrics["_stage"] = stage
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:157]: Good idea. Antoni suggests that users may need this stage information to check their metrics logs.

File: python/ray/train/lightning/_lightning_utils.py [hunk 14]
```diff
@@ -46,11 +50,9 @@ def node_rank(self) -> int:
         return session.get_node_rank()
 
     def set_world_size(self, size: int) -> None:
-        logger.warning("world_size setter is disabled in AIR LightningTrainer.")
```
Comment (gjoliver) [inline python/ray/train/lightning/_lightning_utils.py:49]: sorry to keep you back and forth, getting rid of the warning is ok, can you add a comment to explain a bit why this is disabled instead.
just say this is intentionally disabled, since global_rand() etc return data from Trainer session.

File: python/ray/train/lightning/_lightning_utils.py [hunk 15]
```diff
@@ -94,3 +96,72 @@ def _val_dataloader() -> DataLoader:
...
+    A subclass of ``pytorch_lightning.callbacks.ModelCheckpoint``.
+    This callback function reports the latest metrics to the AIR session and
+    creates an AIR checkpoint whenever a lightning checkpoint is saved.
+    """
+
+    def setup(self, *args, **kwargs) -> None:
+        super().setup(*args, **kwargs)
+        self.is_checkpoint_step = False
+
+    def _session_report(self, trainer: "pl.Trainer", stage: str):
```
Comment (gjoliver) [inline python/ray/train/lightning/_lightning_utils.py:114]: why is metrics reporting part of the checkpoint class?
what if I want to report data / iteration, but don't want to create checkpoints?

File: python/ray/train/lightning/_lightning_utils.py [hunk 16]
```diff
@@ -94,3 +96,72 @@ def _val_dataloader() -> DataLoader:
...
+        """
+
+        # Align the frequency of checkpointing and logging
+        if not self.is_checkpoint_step:
+            return
+
+        # Report latest logged metrics
+        metrics = {"report_on": stage}
+        for k, v in self._monitor_candidates(trainer).items():
+            if k == "report_on":
```
Comment (gjoliver) [inline python/ray/train/lightning/_lightning_utils.py:129]: don't feel particularly safe about this. why do we have this keyword, and it's not even __ prefixed ...
also we may be logging this warning msg every time this is called.
is there a list of such keywords defined somewhere?

File: python/ray/train/lightning/lightning_trainer.py [hunk 17]
```diff
@@ -308,6 +309,17 @@ def __init__(
         preprocessor: Optional[Preprocessor] = None,
         resume_from_checkpoint: Optional[Checkpoint] = None,
     ):
+        run_config = run_config or RunConfig()
+        lightning_config = lightning_config or LightningConfigBuilder().build()
```
Comment (gjoliver) [inline python/ray/train/lightning/lightning_trainer.py:313]: wait, do we want people to pass in a dict or the raw Builder()?
the type annotation says it's a raw dict already.

File: python/ray/train/lightning/lightning_trainer.py [hunk 18]
```diff
@@ -308,6 +309,17 @@ def __init__(
...
+
+        self._check_checkpoint_configs(
+            ptl_ckpt_config=lightning_config["_model_checkpoint_config"],
+            air_ckpt_config=run_config.checkpoint_config,
+        )
+
+        # Disable strict checking to allow metric reporting at different frequencies
+        os.environ["TUNE_DISABLE_STRICT_METRIC_CHECKING"] = "1"
```
Comment (gjoliver) [inline python/ray/train/lightning/lightning_trainer.py:321]: 😓 
any chance we can disable the csv writer so we don't have to do this hack here?

**Review by gjoliver — COMMENTED** [review]
pretty nice. a few questions.

**Review by gjoliver — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]
stamp for docs

**Comment by ollie-iterators** [comment]
This test is causing documentation to fail
